### ISSUE TITLE
Bug: Inconsistent scroll position changes when reordering a list of items with react keys

### ISSUE THREAD

**Issue by natagran** [facebook/react#27044]
When reordering the list of items that have correct React keys assigned (the key is an actual id of an item) we observe the unexpected and inconsistent scroll position changes due to the way React modifies the DOM tree. However, if we remove the React key from the list items altogether, everything works as expected, the scroll position stays the same and there are no unexpected jumps.

I'm attaching the Codesandbox and a screen recording to make it easier to understand the issue.

https://github.com/facebook/react/assets/7542744/1004307a-e176-4cfb-9c3e-891365b824b8

React version: 18.2.0.

## Steps To Reproduce

1. Open and run [this sandbox](https://codesandbox.io/s/red-snowflake-l8d6z9?file=/App.js).
2. Scroll the list a little bit and click the "Reverse" button. Notice that the list gets reordered but the scroll position doesn't change.
3. Scroll the list till the very end and click the "Reverse" button again. Notice that the list gets reordered and the page jumps - the scroll position changes.
4. Remove the key from the `<li>` completely (`App.js`, line 9). Try to scroll the list to the middle/end again and click the "Reverse" button again. Notice that the list gets reordered but the scroll position stays consistent and doesn't change.

Link to code example:

https://codesandbox.io/s/red-snowflake-l8d6z9?file=/App.js

## The current behavior

Page jumps inconsistently when reordering the list items that have correct React keys assigned.

## The expected behavior

When reordering the list items that have the correct React keys assigned, the scroll position should stay consistent and shouldn't change.

**Comment by ANKITSINGH065** [comment]
The key prop in React is used to uniquely identify each item in a list when rendering dynamically. It helps React efficiently update the DOM tree by keeping track of which elements have changed, been added, or been removed.

In your code, if the key prop is set to an actual id of an item, React uses that key to determine the identity of each list item. When you reorder the list by reversing the items array, React recognizes that the order has changed based on the keys and updates the DOM accordingly. However, this can sometimes lead to unexpected scroll position changes due to the way the browser renders and positions elements in the DOM.

By removing the key prop altogether, you effectively disable React's ability to track individual items in the list. As a result, when you reorder the list, React treats each list item as a new element and re-renders them from scratch. Since the elements are rendered in the original order, there are no unexpected scroll position changes.

It's important to note that removing the key prop can have negative consequences in terms of performance and efficient updates. React relies on the key prop to optimize rendering, especially when dealing with large lists. If the order of items is not important in your case and you don't experience any negative side effects, removing the key prop might be an acceptable solution. However, if maintaining the order and preserving the correct scroll position are important, you should explore alternative approaches to address the issue, such as using a more specific key that doesn't change when the list is reordered or finding ways to mitigate the scroll position changes when updating the list.

**Comment by natagran** [comment]
Hi @ANKITSINGH065, thanks for explaining how the React keys work, however, it doesn’t address the issue I’m reporting. I do understand the mechanics and importance of using correct React keys in order to ensure efficient DOM updates. This is why I would like to leverage them when implementing sorting in larger lists, but without compromising the user experience. My example already uses a specific React key that is bound to the item and doesn’t change when the list is reordered, however, the scroll position issue I’m reporting is still present. I see that it’s happening because of the way React updates the DOM after comparing the keys but it doesn’t look like expected behavior.

**Comment by fellmann** [comment]
TL;DR: Add CSS `overflow-anchor: none` to the items.

I also ran into this problem and thought of a React bug. But I found the behavior consistent in Chrome, Firefox, React and svelte, too.

So I searched further and found https://stackoverflow.com/a/54951893/8035079
It seems, scroll anchoring is widely implemented in most browsers as default as of today: https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-anchor

To disable it, you can add `overflow-anchor: none` to the list items. So your example can be fixed by using:
```
    <li key={person.id} style={{ overflowAnchor: "none" }}>
```
